File: quant-black-box/backend/app/routers/mc.py
from fastapi import APIRouter

from .. import engine
from ..schemas import McRequest, McResponse, McStatsModel

router = APIRouter()


@router.post("", response_model=McResponse)
def simulate(req: McRequest) -> McResponse:
    sim = engine.simulate_mc(
        engine.McParams(
            S0=req.S0, mu=req.mu, sig=req.sig, T=req.T, npaths=req.npaths, gam=req.gam
        )
    )
    stats = engine.mc_stats(sim, req.gam)
    return McResponse(
        params=req,
        steps=engine.MC_STEPS,
        N=sim.N,
        term=sim.term,
        paths=sim.sorted if req.include_paths else None,
        stats=McStatsModel(
            mean=stats.mean,
            median=stats.median,
            sd=stats.sd,
            var5=stats.var5,
            max=stats.max,
            min=stats.min,
            losses=stats.losses,
            mlr=stats.mlr,
            util=stats.util,
            ci95=stats.ci95,
        ),
    )

File: quant-black-box/server/src/routes/mc.ts
import { Router } from 'express';
import { z } from 'zod';
import { simulateMc, mcStats } from '../engine.js';

const router = Router();

const mcSchema = z.object({
  S0: z.coerce.number().gt(0),
  mu: z.coerce.number(),
  sig: z.coerce.number().gt(0),
  T: z.coerce.number().gt(0),
  npaths: z.coerce.number().min(1).max(10000),
  gam: z.coerce.number(),
  include_paths: z.coerce.boolean().default(true),
});

router.post('/', (req, res) => {
  const parsed = mcSchema.safeParse(req.body);
  if (!parsed.success) {
    res.status(422).json({ ok: false, error: 'Invalid parameters', details: parsed.error.flatten() });
    return;
  }
  const { S0, mu, sig, T, npaths, gam, include_paths } = parsed.data;
  const sim = simulateMc({ S0, mu, sig, T, npaths, gam });
  const stats = mcStats(sim, gam);
  res.json({
    params: { S0, mu, sig, T, npaths, gam },
    steps: 60,
    N: sim.N,
    term: sim.term,
    paths: include_paths ? sim.sorted : null,
    stats,
  });
});

export default router;
